compiler/src/iree/compiler/Dialect/HAL/Analysis/DeviceAnalysis.h

```diff
@@ -0,0 +1,104 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_DIALECT_HAL_ANALYSIS_DEVICETARGET_H_
+#define IREE_COMPILER_DIALECT_HAL_ANALYSIS_DEVICETARGET_H_
+
+#include <optional>
+
+#include "iree/compiler/Dialect/HAL/Analysis/DeviceSet.h"
+#include "iree/compiler/Dialect/HAL/IR/HALTypes.h"
+#include "iree/compiler/Dialect/Util/Analysis/DFX/Solver.h"
+#include "iree/compiler/Dialect/Util/Analysis/Explorer.h"
+#include "llvm/ADT/DenseSet.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Support/LLVM.h"
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+//===----------------------------------------------------------------------===//
+// DeviceAnalysis
+//===----------------------------------------------------------------------===//
+
+// Performs whole-program analysis of device traits (limits, configuration, etc)
+// and allows for queries against `!hal.device` values for known traits.
+//
+// Though safe to run at any time this may not provide meaningful results until
+// after devices have been materialized and the program has been converted into
+// the HAL dialect.
+class DeviceAnalysis {
```

Comment:
Does this have tests? We have a few cc_tests for compiler code, if a lit_test doesn't make sense: https://github.com/iree-org/iree/tree/main/compiler/src/iree/compiler/Codegen/Common/unittests

Fine to skip tests if you think the code is straightforward enough, the code is covered by other tests indirectly, or unit testing looks too difficult to add.

Reply:
Would be nice to have a debug pass (AnnotateDevicesPass or something similar to the AnnotateAffinitiesPass that comes later in the chain) - then we could have lit tests that cover it. I think by the end of the chain it has full coverage via all the passes that use it, but it would be nice to have a dedicated test and the debugging too. Filed #17942.